File: sdk-node/tests/webhook/shared-fixtures.test.ts
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import {
  decodeRawEmail,
  handleWebhook,
  isRawIncluded,
  PrimitiveWebhookError,
  parseWebhookEvent,
  RawEmailDecodeError,
  safeValidateEmailReceivedEvent,
  validateEmailAuth,
  validateEmailReceivedEvent,
  verifyRawEmailDownload,
  verifyWebhookSignature,
  WebhookValidationError,
  WebhookVerificationError,
} from "../../src/index.js";
import { signWebhookPayload } from "../../src/webhook/signing.js";
import {
  signStandardWebhooksPayload,
  verifyStandardWebhooksSignature,
} from "../../src/webhook/standard-webhooks.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const FIXTURES_DIR = join(__dirname, "../../../test-fixtures");

function loadJson<T>(...parts: string[]): T {
  return JSON.parse(readFileSync(join(FIXTURES_DIR, ...parts), "utf8")) as T;
}

function loadText(...parts: string[]): string {
  return readFileSync(join(FIXTURES_DIR, ...parts), "utf8");
}

describe("shared compatibility fixtures", () => {
  it("validates shared webhook cases", () => {
    const fixtures = loadJson<{
      cases: Array<{
        name: string;
        payload: unknown;
        expected: { valid: boolean; id?: string; error_code?: string };
      }>;
    }>("webhook", "validation-cases.json");

    for (const testCase of fixtures.cases) {
      if (testCase.expected.valid) {
        const event = validateEmailReceivedEvent(testCase.payload);
        expect(event.id, testCase.name).toBe(testCase.expected.id);
        const safeResult = safeValidateEmailReceivedEvent(testCase.payload);
        expect(safeResult.success, testCase.name).toBe(true);
      } else {
        expect(
          () => validateEmailReceivedEvent(testCase.payload),
          testCase.name,
        ).toThrow(WebhookValidationError);
        const safeResult = safeValidateEmailReceivedEvent(testCase.payload);
        expect(safeResult.success, testCase.name).toBe(false);
        if (!safeResult.success) {
          expect(safeResult.error.code, testCase.name).toBe(
            testCase.expected.error_code,
          );
        }
      }
    }
  });

  it("verifies shared signing fixtures", () => {
    const fixtures = loadJson<{
      cases: Array<{
        name: string;
        raw_body: string;
        secret: string;
        timestamp: number;
        verify_secret?: string;
        now_seconds?: number;
        signature_header?: string;
        expected_v1: string;
        expected_valid: boolean;
        expected_error_code?: string;
      }>;
    }>("signing", "vectors.json");

    for (const testCase of fixtures.cases) {
      const signed = signWebhookPayload(
        testCase.raw_body,
        testCase.secret,
        testCase.timestamp,
      );
      expect(signed.v1, testCase.name).toBe(testCase.expected_v1);

      const verifySecret = testCase.verify_secret ?? testCase.secret;
      const nowSeconds = testCase.now_seconds ?? testCase.timestamp;
      const signatureHeader = testCase.signature_header ?? signed.header;

      if (testCase.expected_valid) {
        expect(
          verifyWebhookSignature({
            rawBody: testCase.raw_body,
            signatureHeader,
            secret: verifySecret,
            nowSeconds,
          }),
          testCase.name,
        ).toBe(true);
      } else {
        try {
          verifyWebhookSignature({
            rawBody: testCase.raw_body,
            signatureHeader,
            secret: verifySecret,
            nowSeconds,
          });
          expect.fail(`Expected verification failure for ${testCase.name}`);
        } catch (error) {
          expect(error, testCase.name).toBeInstanceOf(WebhookVerificationError);
          expect((error as WebhookVerificationError).code, testCase.name).toBe(
            testCase.expected_error_code,
          );
        }
      }
    }
  });

  it("classifies shared auth fixtures", () => {
    const fixtures = loadJson<{
      cases: Array<{
        name: string;
        input: Parameters<typeof validateEmailAuth>[0];
        expected: { verdict: string; confidence: string };
      }>;
    }>("auth", "cases.json");

    for (const testCase of fixtures.cases) {
      const result = validateEmailAuth(testCase.input);
      expect(result.verdict, testCase.name).toBe(testCase.expected.verdict);
      expect(result.confidence, testCase.name).toBe(
        testCase.expected.confidence,
      );
    }
  });

  it("handles shared raw content fixtures", () => {
    const fixtures = loadJson<{
      cases: Array<{
        name: string;
        event: unknown;
        download_bytes_utf8?: string;
        expected: {
          included: boolean;
          decoded_utf8?: string;
          decode_error_code?: string;
          verify_download?: boolean;
          verify_download_error_code?: string;
        };
      }>;
    }>("raw", "cases.json");

    for (const testCase of fixtures.cases) {
      expect(isRawIncluded(testCase.event), testCase.name).toBe(
        testCase.expected.included,
      );

      if (testCase.expected.decoded_utf8) {
        expect(
          decodeRawEmail(testCase.event).toString("utf8"),
          testCase.name,
        ).toBe(testCase.expected.decoded_utf8);
      }

      if (testCase.expected.decode_error_code) {
        try {
          decodeRawEmail(testCase.event);
          expect.fail(`Expected decode failure for ${testCase.name}`);
        } catch (error) {
          expect(error, testCase.name).toBeInstanceOf(RawEmailDecodeError);
          expect((error as RawEmailDecodeError).code, testCase.name).toBe(
            testCase.expected.decode_error_code,
          );
        }
      }

      if (testCase.expected.verify_download) {
        const downloaded = Buffer.from(
          testCase.download_bytes_utf8 ?? "",
          "utf8",
        );
        expect(
          verifyRawEmailDownload(downloaded, testCase.event),
          testCase.name,
        ).toEqual(downloaded);
      }

      if (testCase.expected.verify_download_error_code) {
        try {
          verifyRawEmailDownload(
            Buffer.from(testCase.download_bytes_utf8 ?? "", "utf8"),
            testCase.event,
          );
          expect.fail(
            `Expected download verification failure for ${testCase.name}`,
          );
        } catch (error) {
          expect(error, testCase.name).toBeInstanceOf(RawEmailDecodeError);
          expect((error as RawEmailDecodeError).code, testCase.name).toBe(
            testCase.expected.verify_download_error_code,
          );
        }
      }
    }
  });

  it("parses shared webhook event fixtures", () => {
    const fixtures = loadJson<{
      cases: Array<{
        name: string;
        input?: unknown;
        input_fixture?: string[];
        expected: {
          kind: "email.received" | "unknown" | "error";
          event?: string;
          id?: string;
          version?: string;
          error_code?: string;
        };
      }>;
    }>("parse-webhook-event", "cases.json");

    for (const testCase of fixtures.cases) {
      const input = testCase.input_fixture
        ? loadJson(...testCase.input_fixture)
        : testCase.input;

      if (testCase.expected.kind === "error") {
        try {
          parseWebhookEvent(input);
          expect.fail(`Expected parse failure for ${testCase.name}`);
        } catch (error) {
          expect(error, testCase.name).toBeInstanceOf(PrimitiveWebhookError);
          expect((error as PrimitiveWebhookError).code, testCase.name).toBe(
            testCase.expected.error_code,
          );
        }
        continue;
      }

      const event = parseWebhookEvent(input);
      expect(event.event, testCase.name).toBe(
        testCase.expected.event ?? testCase.expected.kind,
      );

      if (testCase.expected.kind === "email.received") {
        expect((event as { id: string }).id, testCase.name).toBe(
          testCase.expected.id,
        );
      } else {
        expect((event as { id?: string }).id, testCase.name).toBe(
          testCase.expected.id,
        );
        expect((event as { version?: string }).version, testCase.name).toBe(
          testCase.expected.version,
        );
      }
    }
  });

  it("handles shared webhook fixtures", () => {
    const fixtures = loadJson<{
      cases: Array<{
        name: string;
        body?: string;
        body_fixture?: string[];
        headers: Record<string, string>;
        secret: string;
        sign_secret?: string;
        timestamp?: number;
        tolerance_seconds?: number;
        expected: {
          valid: boolean;
          id?: string;
          error_code?: string;
        };
      }>;
    }>("handle-webhook", "cases.json");

    for (const testCase of fixtures.cases) {
      const body = testCase.body_fixture
        ? loadText(...testCase.body_fixture)
        : (testCase.body ?? "");
      const signSecret = testCase.sign_secret ?? testCase.secret;
      const signed =
        Object.values(testCase.headers).includes("{signed}") ||
        testCase.timestamp !== undefined
          ? signWebhookPayload(body, signSecret, testCase.timestamp)
          : null;
      const headers = Object.fromEntries(
        Object.entries(testCase.headers).map(([key, value]) => [
          key,
          value === "{signed}" ? (signed?.header ?? "") : value,
        ]),
      );

      if (testCase.expected.valid) {
        const event = handleWebhook({
          body,
          headers,
          secret: testCase.secret,
          toleranceSeconds: testCase.tolerance_seconds,
        });
        expect(event.id, testCase.name).toBe(testCase.expected.id);
        continue;
      }

      try {
        handleWebhook({
          body,
          headers,
          secret: testCase.secret,
          toleranceSeconds: testCase.tolerance_seconds,
        });
        expect.fail(`Expected handleWebhook failure for ${testCase.name}`);
      } catch (error) {
        expect(error, testCase.name).toBeInstanceOf(PrimitiveWebhookError);
        expect((error as PrimitiveWebhookError).code, testCase.name).toBe(
          testCase.expected.error_code,
        );
      }
    }
  });

  it("verifies shared standard webhooks signing fixtures", () => {
    const fixtures = loadJson<{
      cases: Array<{
        name: string;
        raw_body: string;
        secret: string;
        msg_id: string;
        timestamp: number;
        verify_secret?: string;
        now_seconds?: number;
        webhook_signature_header?: string;
        expected_signature: string;
        expected_valid: boolean;
        expected_error_code?: string;
      }>;
    }>("signing", "standard-webhooks-vectors.json");

    for (const testCase of fixtures.cases) {
      const signed = signStandardWebhooksPayload(
        testCase.raw_body,
        testCase.secret,
        testCase.msg_id,
        testCase.timestamp,
      );
      expect(signed.signature, testCase.name).toBe(
        `v1,${testCase.expected_signature}`,
      );

      const verifySecret = testCase.verify_secret ?? testCase.secret;
      const nowSeconds = testCase.now_seconds ?? testCase.timestamp;
      const signatureHeader =
        testCase.webhook_signature_header ?? signed.signature;

      if (testCase.expected_valid) {
        expect(
          verifyStandardWebhooksSignature({
            rawBody: testCase.raw_body,
            msgId: testCase.msg_id,
            timestamp: String(testCase.timestamp),
            signatureHeader,
            secret: verifySecret,
            nowSeconds,
          }),
          testCase.name,
        ).toBe(true);
      } else {
        try {
          verifyStandardWebhooksSignature({
            rawBody: testCase.raw_body,
            msgId: testCase.msg_id,
            timestamp: String(testCase.timestamp),
            signatureHeader,
            secret: verifySecret,
            nowSeconds,
          });
          expect.fail(`Expected verification failure for ${testCase.name}`);
        } catch (error) {
          expect(error, testCase.name).toBeInstanceOf(WebhookVerificationError);
          expect((error as WebhookVerificationError).code, testCase.name).toBe(
            testCase.expected_error_code,
          );
        }
      }
    }
  });

  it("handles shared standard webhooks webhook fixtures", () => {
    const fixtures = loadJson<{
      cases: Array<{
        name: string;
        body?: string;
        body_fixture?: string[];
        headers: Record<string, string>;
        secret: string;
        sign_secret?: string;
        msg_id?: string;
        timestamp?: number;
        tolerance_seconds?: number;
        expected: {
          valid: boolean;
          id?: string;
          error_code?: string;
        };
      }>;
    }>("handle-webhook", "standard-webhooks-cases.json");

    for (const testCase of fixtures.cases) {
      const body = testCase.body_fixture
        ? loadText(...testCase.body_fixture)
        : (testCase.body ?? "");
      const signSecret = testCase.sign_secret ?? testCase.secret;
      const msgId = testCase.msg_id ?? "msg_default";
      const needsSign = Object.values(testCase.headers).includes(
        "{signed_standard}",
      );
      const signed = needsSign
        ? signStandardWebhooksPayload(
            body,
            signSecret,
            msgId,
            testCase.timestamp,
          )
        : null;

      const headers = Object.fromEntries(
        Object.entries(testCase.headers).map(([key, value]) => {
          if (value === "{signed_standard}")
            return [key, signed?.signature ?? ""];
          if (value === "{timestamp}")
            return [key, String(signed?.timestamp ?? testCase.timestamp ?? "")];
          return [key, value];
        }),
      );

      if (testCase.expected.valid) {
        const event = handleWebhook({
          body,
          headers,
          secret: testCase.secret,
          toleranceSeconds: testCase.tolerance_seconds,
        });
        expect(event.id, testCase.name).toBe(testCase.expected.id);
        continue;
      }

      try {
        handleWebhook({
          body,
          headers,
          secret: testCase.secret,
          toleranceSeconds: testCase.tolerance_seconds,
        });
        expect.fail(`Expected handleWebhook failure for ${testCase.name}`);
      } catch (error) {
        expect(error, testCase.name).toBeInstanceOf(PrimitiveWebhookError);
        expect((error as PrimitiveWebhookError).code, testCase.name).toBe(
          testCase.expected.error_code,
        );
      }
    }
  });
});

File: sdk-go/shared_fixtures_test.go
package primitive

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"testing"
)

func loadFixtureCases[T any](t *testing.T, parts ...string) T {
	t.Helper()
	_, filename, _, _ := runtime.Caller(0)
	base := filepath.Join(filepath.Dir(filename), "..", "test-fixtures")
	all := append([]string{base}, parts...)
	data, err := os.ReadFile(filepath.Join(all...))
	if err != nil {
		t.Fatalf("read fixture file: %v", err)
	}
	var value T
	if err := json.Unmarshal(data, &value); err != nil {
		t.Fatalf("decode fixture file: %v", err)
	}
	return value
}

func loadFixtureText(t *testing.T, parts ...string) string {
	t.Helper()
	_, filename, _, _ := runtime.Caller(0)
	base := filepath.Join(filepath.Dir(filename), "..", "test-fixtures")
	all := append([]string{base}, parts...)
	data, err := os.ReadFile(filepath.Join(all...))
	if err != nil {
		t.Fatalf("read fixture file: %v", err)
	}
	return string(data)
}

func webhookErrorCode(err error) (string, bool) {
	var verificationErr *WebhookVerificationError
	if errors.As(err, &verificationErr) {
		return verificationErr.Code(), true
	}
	var payloadErr *WebhookPayloadError
	if errors.As(err, &payloadErr) {
		return payloadErr.Code(), true
	}
	var validationErr *WebhookValidationError
	if errors.As(err, &validationErr) {
		return validationErr.Code(), true
	}
	var decodeErr *RawEmailDecodeError
	if errors.As(err, &decodeErr) {
		return decodeErr.Code(), true
	}
	return "", false
}

func TestSharedCompatibilityFixtures(t *testing.T) {
	t.Run("webhook validation", func(t *testing.T) {
		fixtures := loadFixtureCases[struct {
			Cases []struct {
				Name     string         `json:"name"`
				Payload  map[string]any `json:"payload"`
				Expected struct {
					Valid     bool   `json:"valid"`
					ID        string `json:"id"`
					ErrorCode string `json:"error_code"`
				} `json:"expected"`
			} `json:"cases"`
		}](t, "webhook", "validation-cases.json")

		for _, testCase := range fixtures.Cases {
			if testCase.Expected.Valid {
				event, err := ValidateEmailReceivedEvent(testCase.Payload)
				if err != nil {
					t.Fatalf("%s: ValidateEmailReceivedEvent returned error: %v", testCase.Name, err)
				}
				if event.ID != testCase.Expected.ID {
					t.Fatalf("%s: unexpected event ID %q", testCase.Name, event.ID)
				}
				if !SafeValidateEmailReceivedEvent(testCase.Payload).Success {
					t.Fatalf("%s: expected safe validation success", testCase.Name)
				}
				continue
			}

			_, err := ValidateEmailReceivedEvent(testCase.Payload)
			var validationErr *WebhookValidationError
			if !errors.As(err, &validationErr) {
				t.Fatalf("%s: expected WebhookValidationError, got %v", testCase.Name, err)
			}
			if validationErr.Code() != testCase.Expected.ErrorCode {
				t.Fatalf("%s: unexpected error code %q", testCase.Name, validationErr.Code())
			}
		}
	})

	t.Run("signing", func(t *testing.T) {
		fixtures := loadFixtureCases[struct {
			Cases []struct {
				Name              string `json:"name"`
				RawBody           string `json:"raw_body"`
				Secret            string `json:"secret"`
				Timestamp         int64  `json:"timestamp"`
				VerifySecret      string `json:"verify_secret"`
				NowSeconds        int64  `json:"now_seconds"`
				SignatureHeader   string `json:"signature_header"`
				ExpectedV1        string `json:"expected_v1"`
				ExpectedValid     bool   `json:"expected_valid"`
				ExpectedErrorCode string `json:"expected_error_code"`
			} `json:"cases"`
		}](t, "signing", "vectors.json")

		for _, testCase := range fixtures.Cases {
			signed, err := SignWebhookPayload(testCase.RawBody, testCase.Secret, testCase.Timestamp)
			if err != nil {
				t.Fatalf("%s: SignWebhookPayload returned error: %v", testCase.Name, err)
			}
			if signed.V1 != testCase.ExpectedV1 {
				t.Fatalf("%s: unexpected signature %q", testCase.Name, signed.V1)
			}

			verifySecret := testCase.VerifySecret
			if verifySecret == "" {
				verifySecret = testCase.Secret
			}
			nowSeconds := testCase.NowSeconds
			if nowSeconds == 0 {
				nowSeconds = testCase.Timestamp
			}
			signatureHeader := testCase.SignatureHeader
			if signatureHeader == "" {
				signatureHeader = signed.Header
			}

			_, err = VerifyWebhookSignature(VerifyOptions{
				RawBody:         testCase.RawBody,
				SignatureHeader: signatureHeader,
				Secret:          verifySecret,
				NowSeconds:      &nowSeconds,
			})

			if testCase.ExpectedValid {
				if err != nil {
					t.Fatalf("%s: VerifyWebhookSignature returned error: %v", testCase.Name, err)
				}
				continue
			}

			var verificationErr *WebhookVerificationError
			if !errors.As(err, &verificationErr) {
				t.Fatalf("%s: expected WebhookVerificationError, got %v", testCase.Name, err)
			}
			if verificationErr.Code() != testCase.ExpectedErrorCode {
				t.Fatalf("%s: unexpected error code %q", testCase.Name, verificationErr.Code())
			}
		}
	})

	t.Run("auth", func(t *testing.T) {
		fixtures := loadFixtureCases[struct {
			Cases []struct {
				Name     string `json:"name"`
				Input    any    `json:"input"`
				Expected struct {
					Verdict    AuthVerdict    `json:"verdict"`
					Confidence AuthConfidence `json:"confidence"`
				} `json:"expected"`
			} `json:"cases"`
		}](t, "auth", "cases.json")

		for _, testCase := range fixtures.Cases {
			result, err := ValidateEmailAuth(testCase.Input)
			if err != nil {
				t.Fatalf("%s: ValidateEmailAuth returned error: %v", testCase.Name, err)
			}
			if result.Verdict != testCase.Expected.Verdict || result.Confidence != testCase.Expected.Confidence {
				t.Fatalf("%s: unexpected result %#v", testCase.Name, result)
			}
		}
	})

	t.Run("raw", func(t *testing.T) {
		fixtures := loadFixtureCases[struct {
			Cases []struct {
				Name              string         `json:"name"`
				Event             map[string]any `json:"event"`
				DownloadBytesUTF8 string         `json:"download_bytes_utf8"`
				Expected          struct {
					Included                bool   `json:"included"`
					DecodedUTF8             string `json:"decoded_utf8"`
					DecodeErrorCode         string `json:"decode_error_code"`
					VerifyDownload          bool   `json:"verify_download"`
					VerifyDownloadErrorCode string `json:"verify_download_error_code"`
				} `json:"expected"`
			} `json:"cases"`
		}](t, "raw", "cases.json")

		for _, testCase := range fixtures.Cases {
			included, err := IsRawIncluded(testCase.Event)
			if err != nil {
				t.Fatalf("%s: IsRawIncluded returned error: %v", testCase.Name, err)
			}
			if included != testCase.Expected.Included {
				t.Fatalf("%s: unexpected included flag %v", testCase.Name, included)
			}

			if testCase.Expected.DecodedUTF8 != "" {
				decoded, err := DecodeRawEmail(testCase.Event)
				if err != nil {
					t.Fatalf("%s: DecodeRawEmail returned error: %v", testCase.Name, err)
				}
				if string(decoded) != testCase.Expected.DecodedUTF8 {
					t.Fatalf("%s: unexpected decoded content %q", testCase.Name, string(decoded))
				}
			}

			if testCase.Expected.DecodeErrorCode != "" {
				_, err := DecodeRawEmail(testCase.Event)
				var decodeErr *RawEmailDecodeError
				if !errors.As(err, &decodeErr) {
					t.Fatalf("%s: expected RawEmailDecodeError, got %v", testCase.Name, err)
				}
				if decodeErr.Code() != testCase.Expected.DecodeErrorCode {
					t.Fatalf("%s: unexpected error code %q", testCase.Name, decodeErr.Code())
				}
			}

			if testCase.Expected.VerifyDownload {
				verified, err := VerifyRawEmailDownload([]byte(testCase.DownloadBytesUTF8), testCase.Event)
				if err != nil {
					t.Fatalf("%s: VerifyRawEmailDownload returned error: %v", testCase.Name, err)
				}
				if string(verified) != testCase.DownloadBytesUTF8 {
					t.Fatalf("%s: unexpected verified content %q", testCase.Name, string(verified))
				}
			}

			if testCase.Expected.VerifyDownloadErrorCode != "" {
				_, err := VerifyRawEmailDownload([]byte(testCase.DownloadBytesUTF8), testCase.Event)
				var decodeErr *RawEmailDecodeError
				if !errors.As(err, &decodeErr) {
					t.Fatalf("%s: expected RawEmailDecodeError, got %v", testCase.Name, err)
				}
				if decodeErr.Code() != testCase.Expected.VerifyDownloadErrorCode {
					t.Fatalf("%s: unexpected error code %q", testCase.Name, decodeErr.Code())
				}
			}
		}
	})

	t.Run("parse webhook event", func(t *testing.T) {
		fixtures := loadFixtureCases[struct {
			Cases []struct {
				Name         string   `json:"name"`
				Input        any      `json:"input"`
				InputFixture []string `json:"input_fixture"`
				Expected     struct {
					Kind      string `json:"kind"`
					Event     string `json:"event"`
					ID        string `json:"id"`
					Version   string `json:"version"`
					ErrorCode string `json:"error_code"`
				} `json:"expected"`
			} `json:"cases"`
		}](t, "parse-webhook-event", "cases.json")

		for _, testCase := range fixtures.Cases {
			input := testCase.Input
			if len(testCase.InputFixture) > 0 {
				input = loadFixtureCases[any](t, testCase.InputFixture...)
			}

			if testCase.Expected.Kind == "error" {
				_, err := ParseWebhookEvent(input)
				code, ok := webhookErrorCode(err)
				if !ok {
					t.Fatalf("%s: expected PrimitiveWebhookError, got %v", testCase.Name, err)
				}
				if code != testCase.Expected.ErrorCode {
					t.Fatalf("%s: unexpected error code %q", testCase.Name, code)
				}
				continue
			}

			event, err := ParseWebhookEvent(input)
			if err != nil {
				t.Fatalf("%s: ParseWebhookEvent returned error: %v", testCase.Name, err)
			}
			expectedEvent := testCase.Expected.Event
			if expectedEvent == "" {
				expectedEvent = testCase.Expected.Kind
			}
			if event.GetEvent() != expectedEvent {
				t.Fatalf("%s: unexpected event type %q", testCase.Name, event.GetEvent())
			}

			if testCase.Expected.Kind == string(EventTypeEmailReceived) {
				typed, ok := event.(EmailReceivedEvent)
				if !ok {
					t.Fatalf("%s: expected EmailReceivedEvent, got %T", testCase.Name, event)
				}
				if typed.ID != testCase.Expected.ID {
					t.Fatalf("%s: unexpected event ID %q", testCase.Name, typed.ID)
				}
				continue
			}

			unknown, ok := event.(UnknownEvent)
			if !ok {
				t.Fatalf("%s: expected UnknownEvent, got %T", testCase.Name, event)
			}
			if unknown.ID == nil || *unknown.ID != testCase.Expected.ID {
				t.Fatalf("%s: unexpected unknown event ID %#v", testCase.Name, unknown.ID)
			}
			if unknown.Version == nil || *unknown.Version != testCase.Expected.Version {
				t.Fatalf("%s: unexpected unknown event version %#v", testCase.Name, unknown.Version)
			}
		}
	})

	t.Run("handle webhook", func(t *testing.T) {
		fixtures := loadFixtureCases[struct {
			Cases []struct {
				Name             string            `json:"name"`
				Body             string            `json:"body"`
				BodyFixture      []string          `json:"body_fixture"`
				Headers          map[string]string `json:"headers"`
				Secret           string            `json:"secret"`
				SignSecret       string            `json:"sign_secret"`
				Timestamp        *int64            `json:"timestamp"`
				ToleranceSeconds *int64            `json:"tolerance_seconds"`
				Expected         struct {
					Valid     bool   `json:"valid"`
					ID        string `json:"id"`
					ErrorCode string `json:"error_code"`
				} `json:"expected"`
			} `json:"cases"`
		}](t, "handle-webhook", "cases.json")

		for _, testCase := range fixtures.Cases {
			body := testCase.Body
			if len(testCase.BodyFixture) > 0 {
				body = loadFixtureText(t, testCase.BodyFixture...)
			}
			signSecret := testCase.SignSecret
			if signSecret == "" {
				signSecret = testCase.Secret
			}
			signed, err := SignWebhookPayload(body, signSecret)
			if testCase.Timestamp != nil {
				signed, err = SignWebhookPayload(body, signSecret, *testCase.Timestamp)
			}
			if err != nil {
				t.Fatalf("%s: SignWebhookPayload returned error: %v", testCase.Name, err)
			}

			headers := map[string]string{}
			for key, value := range testCase.Headers {
				if value == "{signed}" {
					headers[key] = signed.Header
				} else {
					headers[key] = value
				}
			}

			if testCase.Expected.Valid {
				event, err := HandleWebhook(HandleWebhookOptions{
					Body:             body,
					Headers:          headers,
					Secret:           testCase.Secret,
					ToleranceSeconds: testCase.ToleranceSeconds,
				})
				if err != nil {
					t.Fatalf("%s: HandleWebhook returned error: %v", testCase.Name, err)
				}
				if event.ID != testCase.Expected.ID {
					t.Fatalf("%s: unexpected event ID %q", testCase.Name, event.ID)
				}
				continue
			}

			_, err = HandleWebhook(HandleWebhookOptions{
				Body:             body,
				Headers:          headers,
				Secret:           testCase.Secret,
				ToleranceSeconds: testCase.ToleranceSeconds,
			})
			code, ok := webhookErrorCode(err)
			if !ok {
				t.Fatalf("%s: expected PrimitiveWebhookError, got %v", testCase.Name, err)
			}
			if code != testCase.Expected.ErrorCode {
				t.Fatalf("%s: unexpected error code %q", testCase.Name, code)
			}
		}
	})

	t.Run("standard webhooks signing", func(t *testing.T) {
		fixtures := loadFixtureCases[struct {
			Cases []struct {
				Name                   string  `json:"name"`
				RawBody                string  `json:"raw_body"`
				Secret                 string  `json:"secret"`
				MsgID                  string  `json:"msg_id"`
				Timestamp              int64   `json:"timestamp"`
				VerifySecret           *string `json:"verify_secret"`
				NowSeconds             *int64  `json:"now_seconds"`
				WebhookSignatureHeader *string `json:"webhook_signature_header"`
				ExpectedSignature      string  `json:"expected_signature"`
				ExpectedValid          bool    `json:"expected_valid"`
				ExpectedErrorCode      string  `json:"expected_error_code"`
			} `json:"cases"`
		}](t, "signing", "standard-webhooks-vectors.json")

		for _, testCase := range fixtures.Cases {
			signed, err := SignStandardWebhooksPayload(testCase.RawBody, testCase.Secret, testCase.MsgID, testCase.Timestamp)
			if err != nil {
				t.Fatalf("%s: SignStandardWebhooksPayload returned error: %v", testCase.Name, err)
			}
			expectedSig := "v1," + testCase.ExpectedSignature
			if signed.Signature != expectedSig {
				t.Fatalf("%s: unexpected signature %q, expected %q", testCase.Name, signed.Signature, expectedSig)
			}

			verifySecret := testCase.Secret
			if testCase.VerifySecret != nil {
				verifySecret = *testCase.VerifySecret
			}
			nowSeconds := testCase.Timestamp
			if testCase.NowSeconds != nil {
				nowSeconds = *testCase.NowSeconds
			}
			signatureHeader := signed.Signature
			if testCase.WebhookSignatureHeader != nil {
				signatureHeader = *testCase.WebhookSignatureHeader
			}

			_, err = VerifyStandardWebhooksSignature(StandardWebhooksVerifyOptions{
				RawBody:         testCase.RawBody,
				MsgID:           testCase.MsgID,
				Timestamp:       fmt.Sprintf("%d", testCase.Timestamp),
				SignatureHeader: signatureHeader,
				Secret:          verifySecret,
				NowSeconds:      &nowSeconds,
			})

			if testCase.ExpectedValid {
				if err != nil {
					t.Fatalf("%s: VerifyStandardWebhooksSignature returned error: %v", testCase.Name, err)
				}
				continue
			}

			var verificationErr *WebhookVerificationError
			if !errors.As(err, &verificationErr) {
				t.Fatalf("%s: expected WebhookVerificationError, got %v", testCase.Name, err)
			}
			if verificationErr.Code() != testCase.ExpectedErrorCode {
				t.Fatalf("%s: unexpected error code %q", testCase.Name, verificationErr.Code())
			}
		}
	})

	t.Run("standard webhooks handle webhook", func(t *testing.T) {
		fixtures := loadFixtureCases[struct {
			Cases []struct {
				Name             string            `json:"name"`
				Body             string            `json:"body"`
				BodyFixture      []string          `json:"body_fixture"`
				Headers          map[string]string `json:"headers"`
				Secret           string            `json:"secret"`
				SignSecret       string            `json:"sign_secret"`
				MsgID            string            `json:"msg_id"`
				Timestamp        *int64            `json:"timestamp"`
				ToleranceSeconds *int64            `json:"tolerance_seconds"`
				Expected         struct {
					Valid     bool   `json:"valid"`
					ID        string `json:"id"`
					ErrorCode string `json:"error_code"`
				} `json:"expected"`
			} `json:"cases"`
		}](t, "handle-webhook", "standard-webhooks-cases.json")

		for _, testCase := range fixtures.Cases {
			body := testCase.Body
			if len(testCase.BodyFixture) > 0 {
				body = loadFixtureText(t, testCase.BodyFixture...)
			}
			signSecret := testCase.SignSecret
			if signSecret == "" {
				signSecret = testCase.Secret
			}
			msgID := testCase.MsgID
			if msgID == "" {
				msgID = "msg_default"
			}

			needsSign := false
			for _, v := range testCase.Headers {
				if v == "{signed_standard}" {
					needsSign = true
					break
				}
			}

			var signed StandardWebhooksSignResult
			if needsSign {
				var ts []int64
				if testCase.Timestamp != nil {
					ts = append(ts, *testCase.Timestamp)
				}
				var err error
				signed, err = SignStandardWebhooksPayload(body, signSecret, msgID, ts...)
				if err != nil {
					t.Fatalf("%s: SignStandardWebhooksPayload returned error: %v", testCase.Name, err)
				}
			}

			headers := map[string]string{}
			for key, value := range testCase.Headers {
				switch value {
				case "{signed_standard}":
					headers[key] = signed.Signature
				case "{timestamp}":
					if needsSign {
						headers[key] = fmt.Sprintf("%d", signed.Timestamp)
					} else if testCase.Timestamp != nil {
						headers[key] = fmt.Sprintf("%d", *testCase.Timestamp)
					}
				default:
					headers[key] = value
				}
			}

			if testCase.Expected.Valid {
				event, err := HandleWebhook(HandleWebhookOptions{
					Body:             body,
					Headers:          headers,
					Secret:           testCase.Secret,
					ToleranceSeconds: testCase.ToleranceSeconds,
				})
				if err != nil {
					t.Fatalf("%s: HandleWebhook returned error: %v", testCase.Name, err)
				}
				if event.ID != testCase.Expected.ID {
					t.Fatalf("%s: unexpected event ID %q", testCase.Name, event.ID)
				}
				continue
			}

			_, err := HandleWebhook(HandleWebhookOptions{
				Body:             body,
				Headers:          headers,
				Secret:           testCase.Secret,
				ToleranceSeconds: testCase.ToleranceSeconds,
			})
			code, ok := webhookErrorCode(err)
			if !ok {
				t.Fatalf("%s: expected PrimitiveWebhookError, got %v", testCase.Name, err)
			}
			if code != testCase.Expected.ErrorCode {
				t.Fatalf("%s: unexpected error code %q", testCase.Name, code)
			}
		}
	})
}
